### DISCUSSION TITLE
Recording lighthouse position data

### DISCUSSION THREAD

**Discussion by Itai51** [bitcraze/discussions#1360] [Q&A] [answered]
Hello Bitcraze community!

I have 2-3 lighthouses set up in my work area, and I would like to record positional data for my Crazyflie 2.1 during the course of a short, manual flight. That is, I open the CF Client, turn on the lighthouses, and fly the Crazyflie using a game controller. Then, I would presumably do something before/during this flight which would allow me to record the drone's (X, Y, Z) position (and preferably orientation as well) as reported by the lighthouses during the flight - perhaps 100 times a second or something like that.

How can I achieve this?

Thank you!

**Comment by knmcguire** [comment, marked as answer]
Hi!

I'm assuming that you went through [the gettting started tutorial](https://www.bitcraze.io/documentation/tutorials/getting-started-with-lighthouse/) of the lighthouse system and that that all works out for you? You are able to fly the Crazyflie with no problem.

There are 3 ways to do logging on the Crazyflie:
1- The CFclient: you can try out the [log blocks tab](https://www.bitcraze.io/documentation/repository/crazyflie-clients-python/master/userguides/userguide_client/logblocks_tab/). You'll need to setup the logging configurations here first: https://www.bitcraze.io/documentation/repository/crazyflie-clients-python/master/userguides/userguide_client/#logging
2- You can write a script with the Crazyflie python library that does this, but then you'll need to also implement the manual controller as well. Here are the API instructions of how to do that: https://www.bitcraze.io/documentation/repository/crazyflie-lib-python/master/user-guides/python_api/#logging
3- You can use the SD card deck for even quicker logging. You can find the info on [the product page](https://www.bitcraze.io/products/micro-sd-card-deck/) under usage.

Yes 100 times per second (10 milliseconds) is the consider the minimum what [the logging structure can handle](https://www.bitcraze.io/documentation/repository/crazyflie-lib-python/master/user-guides/python_api/#variables-and-logging). But that is perhaps in an ideal situation, so if it doesn't go as fast as you want it to go, we advise you to use the micro SDcard deck.

**Reply by knmcguire** [reply]
Also I forgot to add that you can find the available logging variables on https://www.bitcraze.io/documentation/repository/crazyflie-firmware/master/api/logs/.

The ones interesting for you is stateEstimate.X/Y/Z: https://www.bitcraze.io/documentation/repository/crazyflie-firmware/master/api/logs/#stateestimate

**Reply by Itai51** [reply]
Hello!

Thank you so much for your help thus far on this inquiry, and on other inquiries. Since our last correspondence, I have successfully run the CF Client through Python on my Windows 11, and I have also been able to follow [this tutorial](https://www.bitcraze.io/documentation/repository/crazyflie-lib-python/master/user-guides/sbs_connect_log_param/) through step 2a.

Unfortunately, if I have the CF Client running in the background, then the script from the tutorial no longer works. Is there a way to run scripts alongside the CF Client? If not, then what is the workaround?

Thank you!

**Reply by Residualstress** [reply]
> Hello!
> 
> Thank you so much for your help thus far on this inquiry, and on other inquiries. Since our last correspondence, I have successfully run the CF Client through Python on my Windows 11, and I have also been able to follow [this tutorial](https://www.bitcraze.io/documentation/repository/crazyflie-lib-python/master/user-guides/sbs_connect_log_param/) through step 2a.
> 
> Unfortunately, if I have the CF Client running in the background, then the script from the tutorial no longer works. Is there a way to run scripts alongside the CF Client? If not, then what is the workaround?
> 
> Thank you!

Hello, the problem you met was due to the same port, i.e., `radio://0/80/2M/E7E7E7E7E7` , used by your script and CF Client. On my experience, if you want to control the CF by scripts and monitor it by Client on the same time, you'd better use another host to connect with CF physically. For example, you can use Raspberry Pi to communicate with CF via UART. Here is the [tutorial](https://www.bitcraze.io/documentation/repository/crazyflie-lib-python/master/development/uart_communication/). In that way, you can run the script on Raspberry Pi to send commands to CF and connect with it via Client on your computer.

Meanwhile, I'm also curious about if we can achieve simultaneous communication between CF and scripts and Client on a single computer over a wireless connection? Maybe we need to use Bluetooth in addition to [Crazyradio PA](https://www.bitcraze.io/products/crazyradio-pa/), that would be a long story.

**Reply by knmcguire** [reply]
Ah no, it  is not possible to connect with the crazyradio with the cfclient and connect with a script to the same crazyflie. CRTP is set that you can only have 1-to-1 connection. That's a quality of the protocol, so bluetooth and PA together won't work either.

The only way that you can potentially hack it is to perhaps use the peer 2 peer port and send custom packets. That's what I've achieved in the past before. 

So the more easier way is to log the variables in the cflib script in a custom way and store it in a .cvs file. I'm sure there are many python examples on the web to achieve this.

**Reply by Itai51** [reply]
> The ones interesting for you is stateEstimate.X/Y/Z: https://www.bitcraze.io/documentation/repository/crazyflie-firmware/master/api/logs/#stateestimate

Hello again!
I've successfully been able to record stateEstimate.xyz, but a lab member pointed out that this data could be a post-processed value returned from a combionation of lighthouse data, built in IMU data, and applied filters. Is this the case, or is it in fact the raw lighthouse data?

Thank you!

**Reply by knmcguire** [reply]
So stateEstimate.xyz is the indeed the estimate of the position that comes out of the extended kalman filter: https://www.bitcraze.io/documentation/repository/crazyflie-firmware/master/functional-areas/sensor-to-control/state_estimators/#extended-kalman-filter.

The raw lighthouse data is not position, but the angle of the light sweeps hitting the image sensor on the lighthouse deck. You can find the landing page to the technical information of the lighthouse here: https://www.bitcraze.io/documentation/system/positioning/ligthouse-positioning-system/

Technically the raw data is even something more rudimentary (but I'll spare the details), but the light-sweeps is the value that goes into the EKF before it is turned into an position estimate.

You could consider selecting crossing beam method for the lighthouse positioning which calculates a position before sending it to the EKF, that can be read out through lighthouse.x,y,z
